lib/active_record_inherit_assoc.rb

```diff
@@ -29,6 +29,42 @@ def attribute_inheritance_hash
   prepend ActiveRecordInheritAssocPrepend
 end
 
+module ActiveRecordInheritPreloadAssocPrepend
+  def self.prepended(mod)
+    mod.class_eval do
+      alias_method_chain :associated_records_by_owner, :inherit
+    end
+  end
+
+  def filter_associated_records_with_inherit!(owner, associated_records)
+    return unless reflection.options[:inherit]
+
+    associated_records.select! do |record|
+      Array(reflection.options[:inherit]).all? do |obj|
+        record.send(obj) == owner.send(obj)
+      end
+    end
+  end
+
+  if ActiveRecord::VERSION::MAJOR >= 4 && ActiveRecord::VERSION::MINOR >= 1
```

Comment:
@grosser I opted for repeating the arguments because in Ruby 2.0, the fancy &method was failing because key-value pairs are being passed as an array `[key, value]`. Otherwise I would have to do this:

```ruby
def filter_associated_records_with_inherit!(args)
  owner, associated_records = args.first
  ...
end
```